### PR TITLE
fix(logger): decorated class methods cannot access `this`

### LINKED ISSUE #1058: Bug: decorated class methods cannot access `this` in Logger
## Bug description

A long form description of this issue is available in #1056. This is an issue that affects all three Powertools utilities because they all share similar implementation for the decorators.

---

When using Powertools in with a Class and when decorating a method, the decorated method does not have access to the correct value of `this` and so it's not able to call other methods nor read attributes from its same class.

This is due to the fact that the decorator implementation uses an `() => ()` arrow function instead of a regular `function () {}` one. This mixes the value of `this` which then causes the wrong value to be passed.

Additionally there's a missing detail in the documentation of the Class instrumentation that prevents compounds the issue described above.

### Expected Behavior

Decorated methods have access to the correct value of `this` and is able to call methods or access attributes/properties.

### Current Behavior

See "Steps to Reproduce" section.

### Possible Solution

Apply the same changes done to Tracer in #1055 

### Steps to Reproduce

```ts
import {
  LambdaInterface,
  SyncHandler,
  AsyncHandler,
} from "@aws-lambda-powertools/commons";
import { Logger } from "@aws-lambda-powertools/logger";

const logger = new Logger({ serviceName: "serverlessAirline" });

class Lambda implements LambdaInterface {
  greeting: string;

  constructor(message: string) {
    this.greeting = message;
  }

  public getGreeting(): string {
    /* ... */
    return `Hello ${this.greeting}`;
  }

  @logger.injectLambdaContext()
  public handler(_event: any, _context: any, _callback: any): void {
    console.log(this.getGreeting());
  }
}

export const myFunction = new Lambda("World");
export const handler = myFunction.handler.bind(myFunction);
```

The value logged from `console.log(this.getGreeting());` should be `Hello World` but instead is `Hello undefined`:
image

### Environment

* **Powertools version used**: 1.1.0
* **Packaging format (Layers, npm)**: all
* **AWS Lambda function runtime:** all
* **Debugging logs:** N/A

### Related issues, RFCs

#1056

### PR DESCRIPTION
**Pull request by dreamorosi**
## Description of your changes

As described in #1058 the decorator implementation was affected by an issue that prevented the decorated method/class to access other class members as the value of `this` was passed incorrectly.

This PR aims at fixing the issue, creating an unit test case, making minimal changes to the function in the e2e test to verify the change, and update docstrings/docs.

A long form explanation of the issue can be found at #1055, when merged this PR will close #1058. 

### How to verify this change

See checks below the PR, also see the results of the [e2e test run](https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/2855192482).

### Related issues, RFCs

**Issue number:** #1058  

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [x] I have made corresponding changes to the *documentation*
- [x] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published* in downstream module
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

N/A

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by dreamorosi** [comment]
Note to reviewers: don't merge this until #1054 is merged.

File: packages/logger/tests/e2e/sampleRate.decorator.test.FunctionCode.ts
Comment (ijemmy) [inline packages/logger/tests/e2e/sampleRate.decorator.test.FunctionCode.ts:31]: (optional) Calling this `dummyMethod` is a bit misleading.  What about `printLogInAllLevels`?

The reason is that these logs are used for assertion later so they are important.

**Review by ijemmy — COMMENTED** [review]
One small nit.

Also, similar to #1055, we should also update the `examples/cdk/src/*.Decorator.ts` files. They still don't use .bind() method. We can have this in another PR if you prefer (then I'll approve this).

**Review by ijemmy — APPROVED** [review]
Approved.

Discussed with @dreamorosi , we'll go with this to push the fix first.

The example will be updated in a separated PR.

**Review by ijemmy — DISMISSED** [review]

**Review by saragerion — APPROVED** [review]

**Review by ijemmy — APPROVED** [review]

**Pull request merged**
